File: lib/tools/position.ts
import { AccountInfo } from "./account";
import { Currency } from "./forex";

export type Direction = "buy" | "sell";

export interface PositionInfo {
  posCurrency: Currency;
  quoteCurrency: Currency;
  conversion: number;
  openPrice: number;
  quantity: number | null;
  direction: Direction;
  margin: number;
  takeProfit: number | null;
  stopLoss: number | null;
}

export interface SetPosCurrencyPositionAction {
  action: "setPosCurrency";
  currency: Currency;
}

export interface SetQuoteCurrencyPositionAction {
  action: "setQuoteCurrency";
  currency: Currency;
}

export interface SetConversionPositionAction {
  action: "setConversion";
  conversion: number;
}

export interface SetOpenPricePositionAction {
  action: "setOpenPrice";
  openPrice: number;
}

export interface SetQuantityPositionAction {
  action: "setQuantity";
  quantity: number | null;
}

export interface SetDirectionPositionAction {
  action: "setDirection";
  direction: Direction;
}

export interface SetMarginPositionAction {
  action: "setMargin";
  margin: number;
}

export interface SetTakeProfitPositionAction {
  action: "setTakeProfit";
  takeProfit: number | null;
}

export interface SetStopLossPositionAction {
  action: "setStopLoss";
  stopLoss: number | null;
}

export type PositionAction =
  | SetPosCurrencyPositionAction
  | SetQuoteCurrencyPositionAction
  | SetConversionPositionAction
  | SetOpenPricePositionAction
  | SetQuantityPositionAction
  | SetDirectionPositionAction
  | SetMarginPositionAction
  | SetTakeProfitPositionAction
  | SetStopLossPositionAction;

export function positionReducer(
  state: PositionInfo,
  action: PositionAction
): PositionInfo {
  switch (action.action) {
    case "setPosCurrency":
      return { ...state, posCurrency: action.currency };
    case "setQuoteCurrency":
      return { ...state, quoteCurrency: action.currency };
    case "setConversion":
      return { ...state, conversion: action.conversion };
    case "setOpenPrice":
      return { ...state, openPrice: action.openPrice };
    case "setQuantity":
      return { ...state, quantity: action.quantity };
    case "setDirection": {
      let takeProfit = state.takeProfit;
      let stopLoss = state.stopLoss;

      if (state.direction !== action.direction) {
        if (typeof takeProfit === "number") {
          let tp_distance =
            state.direction === "buy"
              ? takeProfit - state.openPrice
              : state.openPrice - takeProfit;
          takeProfit =
            action.direction === "buy"
              ? state.openPrice + tp_distance
              : state.openPrice - tp_distance;
        }

        if (typeof stopLoss === "number") {
          let sl_distance =
            state.direction === "buy"
              ? state.openPrice - stopLoss
              : stopLoss - state.openPrice;
          stopLoss =
            action.direction === "buy"
              ? state.openPrice - sl_distance
              : state.openPrice + sl_distance;
        }
      }

      return { ...state, direction: action.direction, takeProfit, stopLoss };
    }
    case "setMargin":
      return { ...state, margin: action.margin };
    case "setTakeProfit":
      return { ...state, takeProfit: action.takeProfit };
    case "setStopLoss":
      return { ...state, stopLoss: action.stopLoss };
    default:
      console.error("Unrecognized position action", action);
      return state;
  }
}

interface PositionSize {
  /// Funds available under margin risk (in account currency)
  available: number;
  /// Funds available under margin risk (in position currency)
  availablePos: number;
  /// Funds available under margin risk (in quote currency)
  availableQuote: number;
  /// Margin available under margin risk (in account currency)
  margin: number;
  /// Margin available (in position currency)
  marginPos: number;
  /// Margin available (in quote currency)
  marginQuote: number;
  /// Quantity affordable at position price (as units)
  quantity: number;
  /// Optional actual position size margin risk
  actual: null | {
    /// The actual quote cost (in quote currency)
    costQuote: number;
    /// The actual position const (in position currency)
    costPos: number;
    /// The actual position cost (in account currency)
    cost: number;
    /// The account margin required (as a %)
    margin: number;
  };
}

export function computePositionSize(
  account: AccountInfo,
  position: PositionInfo
): PositionSize {
  const p_rate = account.exchangeRates.rates.get(position.posCurrency) || 1;
  const q_rate = position.conversion;
  const available = account.amount * account.marginRisk;
  const availablePos = available * p_rate;
  const availableQuote = availablePos * q_rate;
  const margin = available / (position.margin || 1);
  const marginPos = margin * p_rate;
  const marginQuote = marginPos * q_rate;
  const quantity = marginQuote / position.openPrice;

  const computeActual = (actualQuantity: number) => {
    const costQuote = actualQuantity * position.openPrice * position.margin;
    const costPos = costQuote / q_rate;
    const cost = costPos / p_rate;
    const margin = cost / account.amount;

    return { costQuote, costPos, cost, margin };
  };

  const actual =
    typeof position.quantity === "number"
      ? computeActual(position.quantity)
      : null;

  return {
    available,
    availablePos,
    availableQuote,
    margin,
    marginPos,
    marginQuote,
    quantity,
    actual,
  };
}

interface StopLoss {
  /// Funds available under position risk (in account currency)
  available: number;
  /// Funds available under position risk (in position currency)
  availablePos: number;
  /// Funds available under position risk (in quote currency)
  availableQuote: number;
  /// Specified position size
  quantity: number;
  /// Required stop-loss distance
  distance: number;
  /// Optional actual stop-loss assessment
  actual: null | {
    /// The actual stop-loss distance (in position currency)
    distance: number;
    /// The possible loss
    loss: number;
    /// The actual position risk (as a %)
    risk: number;
  };
}

export function computeStopLoss(
  account: AccountInfo,
  position: PositionInfo,
  quantity: number
): StopLoss {
  const p_rate = account.exchangeRates.rates.get(position.posCurrency) || 1;
  const q_rate = position.conversion;
  const available = account.amount * account.positionRisk;
  const availablePos = available * p_rate;
  const availableQuote = availablePos * q_rate;
  const distance = quantity === 0 ? 0 : availableQuote / quantity;

  const computeActual = (stopLoss: number) => {
    const actualDistance =
      position.direction === "buy"
        ? position.openPrice - stopLoss
        : stopLoss - position.openPrice;
    const loss = (actualDistance * quantity) / (p_rate * q_rate);
    const risk = loss / account.amount;

    return {
      distance: actualDistance,
      loss,
      risk,
    };
  };

  const actual =
    typeof position.stopLoss === "number"
      ? computeActual(position.stopLoss)
      : null;

  return {
    available,
    availablePos,
    availableQuote,
    quantity,
    distance,
    actual,
  };
}

interface StopLossQuantity {
  /// Funds available under position risk (in account currency)
  available: number;
  /// Funds available under position risk (in position currency)
  availablePos: number;
  /// Funds available under positin risk (in quote currency)
  availableQuote: number;
  /// Computed stop loss distance (in position currency)
  stopLossDistance: number;
  /// Amount that can be bought at the given stop loss (as units)
  quantity: number;
  /// Required margin for that amount (in account currency)
  margin: number;
}

export function computedStopLossQuantity(
  account: AccountInfo,
  position: PositionInfo
): StopLossQuantity {
  const stopLossDistance =
    typeof position.stopLoss === "number"
      ? position.direction === "buy"
        ? position.openPrice - position.stopLoss
        : position.stopLoss - position.openPrice
      : 0;

  const p_rate = account.exchangeRates.rates.get(position.posCurrency) || 1;
  const q_rate = position.conversion;
  const available = account.amount * account.positionRisk;
  const availablePos = available * p_rate;
  const availableQuote = availablePos * q_rate;
  const quantity = availableQuote / stopLossDistance;
  const margin =
    (quantity * position.openPrice * position.margin) / (p_rate * q_rate);

  return {
    available,
    availablePos,
    availableQuote,
    stopLossDistance,
    quantity,
    margin,
  };
}

File: src/model/trading/position.rs
use std::{fmt::Display, rc::Rc, str::FromStr};

use yew::Reducible;

use crate::model::currency::Currency;

use super::account::Account;

#[derive(Debug, Copy, Clone, PartialEq)]
pub enum Direction {
    Buy,
    Sell,
}

impl Display for Direction {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Buy => write!(f, "Buy"),
            Self::Sell => write!(f, "Sell"),
        }
    }
}

impl FromStr for Direction {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Buy" => Ok(Self::Buy),
            "Sell" => Ok(Self::Sell),
            _ => Err(()),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Position {
    pub position_currency: Currency,
    pub quote_currency: Currency,
    pub conversion: f64,
    pub open_price: f64,
    pub quantity: Option<f64>,
    pub direction: Direction,
    pub margin: f64,
    pub take_profit: Option<f64>,
    pub stop_loss: Option<f64>,
}

impl Default for Position {
    fn default() -> Self {
        Self {
            position_currency: Currency::GBP,
            quote_currency: Currency::GBP,
            conversion: 1.0,
            open_price: 0.0,
            quantity: None,
            direction: Direction::Buy,
            margin: 0.05,
            take_profit: None,
            stop_loss: None,
        }
    }
}

pub enum PositionAction {
    SetPositionCurrency { currency: Currency },
    SetQuoteCurrency { currency: Currency },
    SetConversion { conversion: f64 },
    SetOpenPrice { price: f64 },
    SetQuantity { quantity: Option<f64> },
    SetDirection { direction: Direction },
    SetMargin { margin: f64 },
    SetTakeProfit { price: Option<f64> },
    SetStopLoss { price: Option<f64> },
}

impl Reducible for Position {
    type Action = PositionAction;

    fn reduce(self: Rc<Self>, action: Self::Action) -> Rc<Self> {
        match action {
            PositionAction::SetPositionCurrency { currency } => Self {
                position_currency: currency,
                ..(*self).clone()
            },

            PositionAction::SetQuoteCurrency { currency } => Self {
                quote_currency: currency,
                ..(*self).clone()
            },

            PositionAction::SetConversion { conversion } => Self {
                conversion,
                ..(*self).clone()
            },

            PositionAction::SetOpenPrice { price } => Self {
                open_price: price,
                ..(*self).clone()
            },

            PositionAction::SetQuantity { quantity } => Self {
                quantity,
                ..(*self).clone()
            },

            PositionAction::SetDirection { direction } => Self {
                direction,
                ..(*self).clone()
            },

            PositionAction::SetMargin { margin } => Self {
                margin,
                ..(*self).clone()
            },

            PositionAction::SetTakeProfit { price } => Self {
                take_profit: price,
                ..(*self).clone()
            },

            PositionAction::SetStopLoss { price } => Self {
                stop_loss: price,
                ..(*self).clone()
            },
        }
        .into()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct PositionSize {
    /// Funds available under margin risk (in account currency)
    pub available: f64,
    /// Funds available under margin risk (in position currency)
    pub available_position: f64,
    /// Funds available under margin risk (in quote currency)
    pub available_quote: f64,
    /// Margin available under margin risk (in account currency)
    pub margin: f64,
    /// Margin available (in position currency)
    pub margin_position: f64,
    /// margin available (in quote currency)
    pub margin_quote: f64,
    /// Quantity affordable at position price (in units)
    pub affordable: f64,
    /// Optional actual position size margin risk
    pub actual: Option<ActualPositionSize>,
}

impl PositionSize {
    pub fn compute(account: &Account, position: &Position) -> Self {
        let p_rate = account
            .exchange_rates
            .rates
            .get(&position.position_currency)
            .copied()
            .unwrap_or(1.0);
        let q_rate = position.conversion;
        let available = account.amount * account.margin_risk;
        let available_position = available * p_rate;
        let available_quote = available_position * q_rate;

        let margin = available
            / if position.margin == 0.0 {
                1.0
            } else {
                position.margin
            };

        let margin_position = margin * p_rate;
        let margin_quote = margin_position * q_rate;

        let affordable = if position.open_price == 0.0 {
            0.0
        } else {
            margin_quote / position.open_price
        };

        Self {
            available,
            available_position,
            available_quote,
            margin,
            margin_position,
            margin_quote,
            affordable,
            actual: position.quantity.map(|quantity| {
                ActualPositionSize::compute(account, position, quantity, p_rate, q_rate)
            }),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ActualPositionSize {
    /// The actual quote cost (in quote currency)
    pub cost_quote: f64,
    /// The actual position cost (in position currency)
    pub cost_position: f64,
    /// The actual position cost (in account currency)
    pub cost: f64,
    /// The account margin required (percent)
    pub margin: f64,
}

impl ActualPositionSize {
    pub fn compute(
        account: &Account,
        position: &Position,
        quantity: f64,
        q_rate: f64,
        p_rate: f64,
    ) -> Self {
        let cost_quote = quantity * position.open_price * position.margin;
        let cost_position = cost_quote / q_rate;
        let cost = cost_position / p_rate;
        let margin = cost / account.amount;

        Self {
            cost_quote,
            cost_position,
            cost,
            margin,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct StopLoss {
    /// Funds available under position risk (in account currency)
    pub available: f64,
    /// Funds available under position risk (in position currency)
    pub available_position: f64,
    /// Funds available under position risk (in quote currency)
    pub available_quote: f64,
    /// Specified position size
    pub quantity: f64,
    /// Required stop-loss distance
    pub distance: f64,
    /// Optional actual stop-loss assessment
    pub actual: Option<ActualStopLoss>,
}

impl StopLoss {
    pub fn compute(account: &Account, position: &Position, quantity: f64) -> Self {
        let p_rate = account
            .exchange_rates
            .rates
            .get(&position.position_currency)
            .copied()
            .unwrap_or(1.0);
        let q_rate = position.conversion;
        let available = account.amount * account.position_risk;
        let available_position = available * p_rate;
        let available_quote = available_position * q_rate;
        let distance = if quantity == 0.0 {
            0.0
        } else {
            available_quote / quantity
        };

        Self {
            available,
            available_position,
            available_quote,
            quantity,
            distance,
            actual: position.stop_loss.map(|stop_loss| {
                ActualStopLoss::compute(account, position, quantity, p_rate, q_rate, stop_loss)
            }),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ActualStopLoss {
    /// The actual stop-loss distance (in position currency)
    pub distance: f64,
    /// The possible loss
    pub loss: f64,
    /// The actual position risk (percent)
    pub risk: f64,
}

impl ActualStopLoss {
    pub fn compute(
        account: &Account,
        position: &Position,
        quantity: f64,
        p_rate: f64,
        q_rate: f64,
        stop_loss: f64,
    ) -> Self {
        let distance = match position.direction {
            Direction::Buy => position.open_price - stop_loss,
            Direction::Sell => stop_loss - position.open_price,
        };

        let loss = (distance * quantity) / (p_rate * q_rate);
        let risk = loss / account.amount;

        Self {
            distance,
            loss,
            risk,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct StopLossQuantity {
    /// Funds available under position risk (in account currency)
    pub available: f64,
    /// Funds available under position risk (in position currency)
    pub available_position: f64,
    /// Funds available under position risk (in quote currency)
    pub available_quote: f64,
    /// Computed stop loss distance (in position currency)
    pub distance: f64,
    /// Amount that can be bought at the given stop loss (in units)
    pub affordable: f64,
    /// Required margin for that amount (in account currency)
    pub margin: f64,
}

impl StopLossQuantity {
    pub fn compute(account: &Account, position: &Position) -> Self {
        let distance = if let Some(stop_loss) = position.stop_loss {
            match position.direction {
                Direction::Buy => position.open_price - stop_loss,
                Direction::Sell => stop_loss - position.open_price,
            }
        } else {
            0.0
        };

        let p_rate = account
            .exchange_rates
            .rates
            .get(&position.position_currency)
            .copied()
            .unwrap_or(1.0);
        let q_rate = position.conversion;
        let available = account.amount * account.position_risk;
        let available_position = available * p_rate;
        let available_quote = available_position * q_rate;
        let affordable = available_quote / distance;
        let margin = (affordable * position.open_price * position.margin) / (p_rate * q_rate);

        Self {
            available,
            available_position,
            available_quote,
            distance,
            affordable,
            margin,
        }
    }
}
